File: 104_bk/21261212/cpp/pro06.cpp
/*
Given an array A[1~n], please write code to find the smallest and largest element with function template form efficiently.
What is the complexity (BigO) of your implementation?
[Note] The basic algorithm use O(2n) comparison.

Ans:

gcc --version

gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Under My Ubuntu 22.04.3 LTS VM, execute "gcc -o pro06 pro06.cpp" to test codes.

The function find_min_max takes an array A of type T, its size n, and two references to variables of type T to store the minimum and maximum elements. 
It initializes the minimum and maximum elements to the first element of the array and then iterates over the remaining elements to update the minimum and maximum values. 
The function template has a time complexity of O(n) 1.

In the main function, we create an integer array A with some values and call the find_min_max function to find its smallest and largest elements. 
The output is then printed on the console.

*/

#include <iostream>
#include <algorithm>

template <typename T>
void find_min_max(T* A, int n, T& min, T& max) {
    min = max = A[0];
    for (int i = 1; i < n; ++i) {
        if (A[i] > max) {
            max = A[i];
        }
        if (A[i] < min) {
            min = A[i];
        }
    }
}

int main() {
    int A[] = {1, 2, 3, 4, 5};
    int n = sizeof(A) / sizeof(A[0]);
    int min, max;
    find_min_max(A, n, min, max);
    std::cout << "Smallest element: " << min << std::endl;
    std::cout << "Largest element: " << max << std::endl;
    return 0;
}

File: 104_bk/21261212/csharp/pro06.cs
/*
	Given an array A[1~n], please write code to find the smallest and largest element with function template form efficiently.
What is the complexity (BigO) of your implementation?
[Note] The basic algorithm use O(2n) comparison.
Ans:

The above code uses a divide-and-conquer approach to find the minimum and maximum elements of an array. It compares elements in pairs, and then compares the smaller element with the current minimum and the larger element with the current maximum. This reduces the number of comparisons to 3n/2 - 2 1.

Therefore, the complexity of this implementation is O(n).

*/

public static void FindMinMax<T>(T[] arr, out T min, out T max) where T : IComparable<T>
{
    if (arr == null || arr.Length == 0)
    {
        throw new ArgumentException("The array is null or empty.");
    }

    min = arr[0];
    max = arr[0];

    for (int i = 1; i < arr.Length; i += 2)
    {
        if (i == arr.Length - 1)
        {
            if (arr[i].CompareTo(min) < 0)
            {
                min = arr[i];
            }
            else if (arr[i].CompareTo(max) > 0)
            {
                max = arr[i];
            }
        }
        else
        {
            if (arr[i].CompareTo(arr[i + 1]) < 0)
            {
                if (arr[i].CompareTo(min) < 0)
                {
                    min = arr[i];
                }

                if (arr[i + 1].CompareTo(max) > 0)
                {
                    max = arr[i + 1];
                }
            }
            else
            {
                if (arr[i + 1].CompareTo(min) < 0)
                {
                    min = arr[i + 1];
                }

                if (arr[i].CompareTo(max) > 0)
                {
                    max = arr[i];
                }
            }
        }
    }
}